lib/rrule/rule.rb

```diff
@@ -94,39 +79,61 @@ def all_until(end_date: nil, limit: nil)
 
         results_with_time = generator.combine_dates_and_times(possible_days_of_year, timeset)
         results_with_time.sort.each do |this_result|
-          if end_date
-            if this_result > end_date
-              return result
-            end
-          end
-
           if options[:until]
             if this_result > options[:until]
-              return result
+              raise StopIteration
             end
-            result.push(this_result)
+            yield this_result
           elsif this_result >= dtstart
             total += 1
             if options[:count]
               count -= 1
-              result.push(this_result)
-              return result if count == 0
+              if count < 0
+                raise StopIteration
+              else
+                yield this_result
+              end
             else
-              result.push(this_result)
+              yield this_result
```

Comment:
Rather than three `yield`s, I wonder if this could be simplified to just have one yield? Not familiar enough though really. Seems surprising we check that `this_result >= dtstart` only if there is not an `until`?